Docs 6366 shared resources v4.2 (#1272)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/shared-resources.adoc

endif::[]
:keywords: anypoint studio, shared resources, domains, multiple applications, share ports, domain project

> You can define connector configurations as shared resources to expose them to all applications deployed under the same domain. To use shared resources, you create a Mule domain project and then reference it from each of the projects that use the shared resources defined in the domain. After defining the domain, any Mule app associated with the domain can access resources in this file. +

Reviewer: I think we might be able to get rid of a few words to simplify this, but please do double-check my thinking:
```suggestion
You can define connector configurations to be shared among all applications deployed in one domain. To do so, create a Mule domain project and then reference it from each application. As a result, each app now associated with the Mule domain project can access resources in the domain. +
```
I was a little confused by the original text about apps vs. Mule apps, apps deployed in vs. associated with a domain, domain vs. Mule domain project, apps sharing resources vs. projects using shared resources, and the Mule domain project vs. "this file."
It seemed like the same or similar terms were being used in different ways throughout.
My suggestion might be wrong, but be careful, if you reword this, to be very specific about the differences in these term pairs.

> Note that Mule apps can be associated with only one domain at a time.

Reviewer: ```suggestion
Note that Mule apps are associated with only one domain at a time.
```

> Shared resources allow multiple development teams to work in parallel using the same set of reusable connector configurations. Defining these configurations as shared resources at the domain level allows the team to:

Reviewer: ```suggestion
Sharing resources allows multiple development teams to work in parallel using the same set of connector configurations, enabling the teams to:
```

== Prerequisites

* Anypoint Studio 7.x
* Mule runtime engine 4.x

== Limitations

Defining flows, subflows, or any message processors as shared resources is not supported. Domains are not meant to share behavior. They are meant to share resources only.

== Architecture

> A Mule domain Studio project has the following key files:

Reviewer: Is this Mule domain Studio project the same thing referred to earlier as the Mule domain project/these files?
Should we mention Anypoint Studio in the earlier text?

[%header%autowidth.spread,cols="a,a"]
|===

> |File name |Description

Reviewer: Column headings use title-style capitalization:
```suggestion
|File Name |Description
```

|`/src/main/mule/mule-domain-config.xml` |This is the shared resources configuration file. This file must have the `mule-domain-config.xml` file name.
|`pom.xml` |This is the dependencies file descriptor. It is a Maven POM file with required plugins for domains projects.
|`/mule-artifact.json` |This file is a descriptor for the domain artifact.
|===

> When you package the domain into a JAR file, you can find the following files:

Reviewer: Is packaging the domain into a JAR file what creates the Mule domain project?
Again, I am wondering if the earlier text should specify this: "...you create a Mule domain project..." > "you use Anypoint Studio to package the domain into a JAR file, creating a Mule domain project..."?

[%header%autowidth.spread,cols="a,a"]
|===

> |File name |Description

Reviewer: ditto 
```suggestion
|File Name |Description
```

== Configuration

To use shared resources in your apps, complete the following tasks:

. <<create_domain_project, Create a new domain>>.
. <<define-shared-resources, Define one or more shared resources in that domain>>.
. <<assoc_apps_domain, Associate applications with the domain>>.
. <<reference-shared-resources, Reference shared resources>> in your applications.
. <<deploy-shared-resources, Deploy the domain and the applications with shared resources>>. +

> // Note that you can also deploy domain and apps together in a <<Deploying Domain Bundles, domain bundle>>.

Reviewer: If you do this do you need the Mule domain project?

</mule-domain>
----

> Any Mule application associated with the domain can make use of the shared resources by referencing them within the configuration, the same way you reference a resource within the project itself. In the example below, the HTTP listener connector references the shared resource named `HTTP_Listener_Configuration`.

Reviewer: ```suggestion
Any Mule application associated with the domain can make use of the shared resources by referencing them within the configuration in the same way that you reference a resource within the project itself. In the example that follows, the HTTP Listener connector references the shared resource named `HTTP_Listener_Configuration`.
```

[NOTE]
Mule domain projects cannot be installed using Runtime Manager.

=== Deploy from Anypoint Studio

The following steps describe how to deploy your domain project and the associated applications from Studio:

> . Open the drop-down menu next to the play button and select *Run Configurations*.

Reviewer: Is it clear which button is the play button? Is it labeled Play? Or does it have an icon on it?

Reviewer: And just ignore me if the image makes this clear!

+
image::mruntime-run-configuration.png[run+configuration]
+

> . Select the *General* tab, and tick or untick the boxes next to the projects that you want to always deploy together with the application that is currently selected on the navigation menu to the right.

Reviewer: ```suggestion
. Select the *General* tab and then select the boxes next to the projects that you want to always deploy together with the application that is currently selected in the navigation menu to the right.
```
You would not unselect to always deploy together.... I think mentioning only selecting is clear, but if you like, you can add a phrase about unselecting the projects that you want to omit from the group that is always deployed together.

+
image::mruntime-run-configuration-properties.png[run+configuration+properties]

> === Deploy outside Anypoint Studio

Reviewer: ```suggestion
=== Deploy Outside Anypoint Studio
```

The following steps describe how to deploy your domain project and the applications outside Studio, to a standalone Mule runtime engine:

. In Studio, select *File* > *Export*.

> . In the window that opens, select *Mule* > *Anypoint Studio Project to Mule Deployable Archive (includes Studio metadata)*. +

Reviewer: You probably do not need "In the window that opens".

> This process creates a `.jar` file that you can deploy to a standalone Mule runtime engine.

Reviewer: ```suggestion
This process creates a JAR file that you can deploy to a standalone Mule runtime engine.
```
See style ref. entries about file types vs. file names.

+
image::mruntime-export-studio.png[]
+
. Copy the exported domain's JAR file to `MULE_HOME/domains`.

> . Export each application that you want to deploy with the associated domain.

Reviewer: Export each application or each application's JAR files, as in the next step?

. Copy the exported JAR files for each application that references the domain into the `MULE_HOME/apps` folder.
. Start Mule via the command console.
+
When Mule starts, it first deploys any domains found in the `MULE_HOME/domains` folder, then it deploys the applications in the `MULE_HOME/apps` folder, so that all domains are fully started before the applications start.

////
Need to update this section to Mule 4
=== Deploying Domain Bundles

> You also have the option of bundling the applications associated with a domain in your domain folder, then deploying the entire folder as a bundled unit. To do this, include an `apps` folder in your domain folder structure and place the JAR files of your applications there.

Reviewer: ```suggestion
You can also bundle in your domain folder all of the applications associated with a domain and then deploy the entire folder. To do this, include an `apps` folder in your domain folder structure and place the JAR files of your applications there.
```

image::domainbundle.png[domainBundle]

> The deployment behavior is the same as deploying a domain and apps separately: Mule first deploys the domain, and then the applications. Deploying domain bundles simplifies the deployment mechanism for teams by removing the manual step of deploying applications separately.

Reviewer: ```suggestion
The deployment behavior is the same as when a domain and apps are deployed separately: Mule first deploys the domain and then the applications. Deploying domain bundles simplifies the deployment mechanism for teams by removing the manual step of deploying applications separately.
```

////

> === Deploying Domains using Mule Maven Plugin

Reviewer: ```suggestion
=== Deploying Domains Using Mule Maven Plugin
```
